Refactor the GameProfileManager and make GameProfile immutable.

Cybermaxke opened the pull request:
**API** | [Common](https://github.com/SpongePowered/SpongeCommon/pull/3153)

This PR refactors the `GameProfile`, `GameProfileManager` and `GameProfileCache`. 

- It's no longer possible to add entries to the `GameProfileCache`, only remove them to invalidate.
- `GameProfileCache` is now just a cache, it was previously also a lookup service.
- `GameProfileCache` is no longer replaceable.
- `GameProfile` is now immutable, but has transformation methods.
- "filled" profiles are now just "profile", and non filled ones are "basic profile", e.g. `getProfile` and `getBasicProfile`.
- "filled" was previously a bit confusing, because `isFilled` meant that the profile has a name, although filling a profile through `GameProfileManager` also added properties. This is now renamed to `hasName`.

Cybermaxke committed:
Refactor the GameProfileManager and make GameProfile immutable.

Cybermaxke committed:
Small tweaks.

dualspiral commented on src/main/java/org/spongepowered/api/profile/GameProfileProvider.java:
```diff
@@ -0,0 +1,201 @@
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A service which handles how game profile data is requested.
```
It's not a service.

dualspiral reviewed (approved):
Barring my little javadocs comment, this looks fine.

Cybermaxke committed:
Fix javadoc.

Cybermaxke committed:
Offering no longer resolves game profile properties.

Auto resolving is problematic because it can potentially block the main thread. So it now needs to be done explicitly.

Cybermaxke committed:
Make ProfileNotFoundException a RuntimeException.

Zidane commented on src/main/java/org/spongepowered/api/profile/GameProfileManager.java:
```diff
@@ -24,201 +24,30 @@
-     */
-    GameProfileCache getDefaultCache();
-
+    GameProfileProvider uncached();
```
getUncached?

getDynamic?

With that said, I am not sure how wise it is to expose a provider that doesn't cache. As only due to the fact that profiles can be slow to query and Mojang does rate limit an IP from querying so many per minute I think.

Cybermaxke replied:
The uncached one can still cache to consider rate limits, it just bypasses the default cache which retains profiles for a month.

Zidane commented:
Manually merged, thanks!
